DM-23302: Add ci_cpp package to allow calibration product generation to be tested

czwa opened the pull request:
Make ci_cpp_gen2 functional.

czwa committed:
Update for crosstalk.

czwa committed:
Initial rework.

 - Ingests raw data
 - Ingests LATISS curated calibs
 - Generates bias, dark, flat
 - Processes science data through defect, bias, dark, flat
 - Runs crosstalk measurement code.
 - Functional execution and dependencies.
 - Add config to set visitKey.
 - Use num_process correctly.

czwa committed:
Add semi-functional unit tests for outputs.

Fix tests, make it sconable.

czwa committed:
Simplify SConstruct.

Update and clarify tests.

plazas commented on SConstruct:
```diff
@@ -62,124 +73,149 @@ def runConstructCalib(typeString, priorStage, visitList,
+env.Alias("all", everything)
+Default(everything)
+
+env.Clean(everything, [y for x in everything for y in x] + ['DATA'])
```
If I understood well from the CCP standing meeting this morning, you are going to move most of the code here to call the tests to python, and then have Scons call those python scripts, correct? 
With respect to the LATISS data, I think you said you chose it because you wanted real data for the ci tests, correct? or are there any other reasons?  Or could we use some other data set that may be better behaved, such as BOT? Or even simulations?  If the tests fail, how do we know that it is not because of less-than-ideal data as opposed to mistakes in the code? 
If we wanted to change LATISS dataset in the future, would we need to change all the hard coded places in this code?

plazas replied:
Some of the data in the next runs that will be taken at the BOT, for example, will have datasets based on DMTN-101.

czwa replied:
At the time I had the idea for this, I think LATISS was the only data we had on disk (I may have been mistaken).  I chose a date that looked to have sufficient data to test most things, and set it up as the `testdata_latiss_cpp` package.  One of the suggestions that came during a sprint planning was that it would be good if the testdata is modular, although I never fully implemented it here.  At the moment, only the `TESTDATA_ROOT` value needs to be changed in the SConstruct file to point to a different product dir to change the input.  This was also the reason I moved all the visit/exposure information to the manifest.yaml file in `testdata_latiss_cpp`: so the ci and testing code is general, and all data selection choices are associated with the testdata package.

If we can get BOT data that contains everything, that would be good.  The other reason to choose LATISS is that the single chip is faster to process.  I suspect we can still do that with BOT data, by correctly restricting the dataIds.

plazas replied:
" At the moment, only the TESTDATA_ROOT value needs to be changed in the SConstruct file to point to a different product dir to change the input.":  I still see some specific data-id in the tests and some specific readout noise value too (I think): 2020012800014. and 0.71. Wouldn't these need to be changed too, or generalized?

czwa replied:
Ah, that's a good catch.  Those will need to be generalized as well.

mfisherlevine replied:
I don't think it's fair to say that LATISS data is not good, nor that it's any better behaved than BOT data. In many ways it's quite the opposite - at least LATISS is "nominally" finished and therefore somewhat static (though admittedly right now a little more static than anyone is happy with!).

I think the main argument for doing things with BOT data is because of it being multichip 0 some form of testing that calibration outputs don't collide, for example, or are correctly mapped to detectors when retrieved, and can be retrieved by all logically coherent combinations of detector ID, raft name/detector name etc would be good. Hopefully we can test that in a way that's robust, but only using 2 chips, otherwise it's going to be overly slow and big.

plazas commented on bin.src/dummy.sh:
```diff
@@ -1,3 +1,4 @@
 
-echo "Dummy!"
+echo "A dummy script to likely be replaced later."
 
```
What's the point of having this "dummy" script instead of just not having anything?

czwa replied:
Almost certainly can be removed.

plazas commented on tests/test_bias.py:
```diff
@@ -0,0 +1,182 @@
+        results = self.isrTask.runDataRef(self.dataRef)
+        self.exposure = results.outputExposure
+
+    def test_canary(self):
```
Why is this called "canary"?

plazas replied:
And why do you need to check for value changes?

czwa replied:
I originally put this in to detect algorithm changes, but I suspect this will end up being overly restrictive.  Removed throughout.

plazas commented on tests/test_bias.py:
```diff
@@ -0,0 +1,182 @@
+        error
+        """
+        mean = afwMath.makeStatistics(self.exposure.getImage(), afwMath.MEAN).getValue()
+        print("4.2", mean)
```
Remove this print statement. There's a couple of more below.

czwa replied:
Converted to msg arguments to the asserts (here and elsewhere), so a helpful message hopefully comes out if we fail a test.

plazas commented on tests/test_dark.py:
```diff
@@ -0,0 +1,224 @@
+            print("5.4", amp.getName(), sigma, sigmaClip, np.abs(sigma - sigmaClip)/sigmaClip)
+            self.assertLess(np.abs(sigma - sigmaClip)/sigmaClip, 3.0)
+
+    def test_55(self):
```
Why 150? The beginning of Sec. 5 of DMTN-1010 mentions 100 dark exposures. For the LATISS data that you are using, I'm guessing that there are not enough darks to do this test, correct? If so, ow will it be implemented in the future? With different data?

plazas replied:
(same comment for test_56 and test_57, and for the tests in "test_flat.py" below, which are the same)

czwa replied:
All of the notes are taken directly from the DMTN-101 text.  This one is on page 4 of the 2018-11-26 draft.  In any case, I'm removing tests that aren't implementable at the moment, so this will go away.

plazas commented on tests/test_dark.py:
```diff
@@ -0,0 +1,224 @@
+
+        """
+        pass
+
```
The end of section 5 of DMTN-101 also talks about repeating the tests using Nd/5, 5*Td dark exposures. Those tests are not necessary here?

czwa replied:
I missed that because it wasn't part of the enumerate list.  In any case, as above, I'm removing these tests.

plazas commented on tests/test_defects.py:
```diff
@@ -0,0 +1,151 @@
+        self.config.doUseAtmosphereTransmission = False
+
+        self.isrTask = ipIsr.IsrTask(config=self.config)
+        # This is not an independent frame.
```
What does this mean? ("this is not an independent frame")

czwa replied:
This is one of the frames that went into the construction of the defects.  The majority of the test descriptions want the test performed on frames that were not used in the construction (difficult here due to data restrictions).  I'm not sure it matters here.

mfisherlevine replied:
Ah, I see my comments above came up here already. I stand by my first one though about maybe making this a ticket, which we can then consider.

plazas commented on tests/test_defects.py:
```diff
@@ -0,0 +1,151 @@
+        self.isrTask = ipIsr.IsrTask(config=self.config)
+        # This is not an independent frame.
+        self.dataRef = butler.dataRef('raw', dataId={'detector': 0, 'expId': 2020012800028})
+        results = self.isrTask.runDataRef(self.dataRef)
```
In the notes, it says "Divide each resulting master frame by its median.". Is this happening here?

czwa replied:
Rereading DMTN-101 section 3 suggests that the listed steps describe how the initial defect map should be generated, and not actual tests on the existing defect maps.  The next test attempts to do this, although the proposed method is questionable (an exposure divided by the median must then have a median equal to 1.0).  I've tried to resolve this a bit, but the test likely needs revisiting (outside the scope of this ticket).

plazas commented on tests/test_defects.py:
```diff
@@ -0,0 +1,151 @@
+        self.assertLess(np.abs(mean/median - 1.0), sigma)
+
+    def test_masterFrameSigma(self):
+        """Ill defined.
+
+        Notes
+        -----
+        DMTN-101 3.3
+
+        For each master frame, divide by the standard deviation image,
+        and confirm that the 5-sigma clipped standard is 1.0 to within
+        statistical noise (after masking known defects)
+        """
+        pass
+
+    def test_pixelConsistency(self):
+        """Ill-defined.
+
+        Notes
+        -----
+        DMTN-101 3.4
+
+        for each pixel, check if all master frames are consistent to
+        within statistical noise; if not, confirm that they are
+        included in the defect map.  Also identify pixels in the
+        defect map that are not identified.
+
+        """
+        pass
```
Why are they ill-defined and, in that case, are there plans to propose well-defined tests and update DMTN-101?

czwa replied:
I've deleted the ill defined tests, and will assume a future ticket will add in tests that are defined better.

plazas commented on tests/test_defects.py:
```diff
@@ -0,0 +1,151 @@
+        pass
+
+    def test_darkPixels(self):
+        """Missing data.
```
Will we get this type of data (pocket-pump flats) for the LATISS data that you are using in order to implement this test? In general, are these "pocket-pumped" flats an official calibration product that the Camera will deliver to us?

mfisherlevine replied:
I believe we should be able to take these once we're back online on the mountain. It was certainly discussed as it being needed to be a capability, so let's assume it is for now.

plazas commented on tests/test_flat.py:
```diff
@@ -0,0 +1,224 @@
+from lsst.utils import getPackageDir
+
+
+class DarkTestCases(lsst.utils.tests.TestCase):
```
It seems that you copied the tests in "test_dark.py" here into "test_flat.py". However, there are still some names that should be changed (for example, this class is called "DarkTestCases"). also, the docstrings reference the tests in section 5 of DMTN-101 which are for darks; perhaps there should be an explanation that his is the case.

czwa replied:
No flat tests are defined in DMTN-101.  I thought that I had converted this file, but it appears I did not.  I have switched the test exposure to a flat, so the class will generate a flat-flattened-flat.  We can check the mean and sigma values are plausible.  I've deleted the other tests that are read-noise based (and so not applicable to the flat).

plazas commented on tests/test_linearity.py:
```diff
@@ -0,0 +1,93 @@
+import lsst.utils.tests
+
+
+class PtcTestCases(lsst.utils.tests.TestCase):
```
I have the same question as above about "ill-defined", but, more generally, what are the plans for all these tests (including the ones in the other files) that say "pass"? Is there a time frame to implement these? Also, in DMTN-101 many of the tests specify that the data should be taken in a specific way: for example, for the non-linearity tests. I guess this does not have to necessarily be enforced for these tests, correct?

czwa replied:
I would like to have useful tests to know that the linearity is constructed correctly.  However, those tests have not been defined in a generic way.  If this package existed before now, we likely would have been adding those tests as the `cp_pipe` construction code was added.  I'm adding ticket DM-26396 that can be done in the future to add/update tests.  I am also adding reference to that ticket in all of the test_ scripts, so there is a note pointing to that future work.

mfisherlevine commented on config/constructCalib.py:
```diff
@@ -0,0 +1 @@
+config.visitKeys = ['expId']  # noqa: F821
```
Why does this live here and not in obs_lsst?

mfisherlevine replied:
Or conversely, if it's in obs_lsst, why is it necessary here?

timj replied:
Let alone something called exposure ID being assigned to something called visit...

czwa replied:
I don't know the logic behind this not being set in the LATISS config files.  I was hesitant to try and "fix" it in obs_lsst without knowing how bias/dark/flat generation is being run for the processing repo.  For this dataset, working with expId values is easier.

mfisherlevine commented on tests/test_bias.py:
```diff
@@ -0,0 +1,168 @@
+        self.config.doUseAtmosphereTransmission = False
+
+        self.isrTask = ipIsr.IsrTask(config=self.config)
+        # This is not an independent frame.
```
This might warrant a proper `TODO:` and a ticket number, given that having some reserved biases etc means a new section of the manifest I assume, and retrieving that so that the line below isn't hard coding the visit.

czwa replied:
Adding reference to DM-26396, which I'm using as the catch all for "fix ci_cpp tests" issues.  I had also imagined adding an entry to the manifest: "reservedBias" or something similar.

mfisherlevine replied:
Sounds perfect.

mfisherlevine commented on tests/test_bias.py:
```diff
@@ -0,0 +1,168 @@
+        error
+        """
+        mean = afwMath.makeStatistics(self.exposure.getImage(), afwMath.MEAN).getValue()
+        self.assertLess(np.abs(mean), 1, msg=f"Test 4.2: {mean}")
```
I guess if the tech note says 1 is a magic value then it's maybe legit to hard code it here, but I do wonder. It's using an unclipped mean though, so I can imagine a very bright column or few throwing that out, for example. On the other hand, we're also testing defect finding and application, so maybe that's exactly what we want... Something to think on though.

mfisherlevine replied:
(that's just one of many failure modes that spring to mind though)

czwa replied:
I think something needs to be done to update/finalize DMTN-101, but I'd like to worry about that after the gen3 conversion is finished.

mfisherlevine replied:
OK, fair enough. I'm not sure that totally addresses the hard-coding of this value, but I guess it gives us a chance to revisit all the tests and their threshold values, so yeah, let's resolve for now 👍

mfisherlevine commented on tests/test_bias.py:
```diff
@@ -0,0 +1,168 @@
+                                           afwMath.STDEVCLIP, statControl).getValue()
+            # needs to be < 0.05
+            fractionalError = np.abs(sigma - amp.getReadNoise())/amp.getReadNoise()
+            self.assertLess(fractionalError, 0.71, msg=f"Test 4.3: {amp.getName()} {fractionalError}")
```
Why is that 0.71?

mfisherlevine replied:
Also, assertion message could be a little more verbose, though I admit it has enough detail to find the triggering code, so I'm not too bothered.

czwa replied:
The error value that was triggering previously is 0.70\d\d\d, so I simply rounded up for the initial set of tests.

mfisherlevine replied:
Yeah, it wasn't so much the 0.71 itself, if was the fact it's written right there that it needed to be <0.05, and so I assumed that was the value that would be being tested for.

I totally understand the temporary (I hope) need to have values different to the real ones we want to aim for so that things can pass and we can merge, but I wanted to understand that that was indeed what's happening. Is that the case?

czwa replied:
Yes.  I'd rather get an imperfect ci_cpp available and usable now, and go back and debug why we're not hitting acceptance targets later (with an implied assumption that a better dataset will likely resolve many of these issues).

mfisherlevine commented on tests/test_bias.py:
```diff
@@ -0,0 +1,168 @@
+        -----
+        DMTN-101 4.4
+
+        Run a CR rejection on the result and confirm that the
+        unclipped standard deviation is consistent with the 5-sigma
+        clipped value.
```
This might be a comment for the technote rather than this code, but consistent at what level?

czwa replied:
Yes.

mfisherlevine replied:
😄

mfisherlevine commented on tests/test_bias.py:
```diff
@@ -0,0 +1,168 @@
+
+            # needs to be < 0.05
+            fractionalError = np.abs(sigma - sigmaClip)/sigmaClip
+            self.assertLess(fractionalError, 3.0, msg="Test 4.4: {amp.getName()} {fractionalError}")
```
3?!

czwa replied:
I believe there's a single amp that triggers this.

mfisherlevine replied:
Hmmm... That seems quite bad, but I guess this is a question of test data at some level too, perhaps, plus we've already got something above to review all the various threshold values, so let's file this under that too.

Maybe a ticket to capture the work?

Also, what do you think about pulling these out into at least a bunch of all-caps const-style VARIABLES that live at the top of the file so that we can at least see all the test values in one place, and not have them hiding as hard-coded values in each test? Or do you think that will be more messy, and just the fact that they live in the asserts in each test makes finding them easy enough?

czwa replied:
I've been searching them by finding the assert.  I was assuming DM-26396 will be the ticket that all the test fixes live on.  I've also thought up a possibly over-designed system of putting the test values in the testdata package (probably in yaml as well), with an override parameter for known failures.  Then the logic is that a test with `override=False` must meet the DMTN limits.  However, one with `override=True` can fail the DMTN limit as long as it meets the testdata's limit.  This separates "validation and verification" from "continuous integration" tests.

mfisherlevine replied:
OK, that certainly sounds interesting, and (unsurprisingly) like you're well ahead of the game here 🙂

mfisherlevine commented on tests/test_brighterFatter.py:
```diff
@@ -0,0 +1,50 @@
+        DMTN-101 14.1
+        "Characterize".
+        """
+        pass
```
How are you planning to do this, out of curiosity? Hard code a kernel in here, and just assert they're the same?

mfisherlevine replied:
However though, maybe worth adding a TODO: and a ticket as, I know if this were me, I'd forget to come back and make that.

czwa replied:
DM-26396 referenced now.  Ideally we'd match against the kernel we believe is correct.

mfisherlevine commented on tests/test_dark.py:
```diff
@@ -0,0 +1,170 @@
+        self.config.doUseAtmosphereTransmission = False
+
+        self.isrTask = ipIsr.IsrTask(config=self.config)
+        # This is not an independent frame.
```
Same as previous comment on hardcoded expId and editing the manifest to have a per-reserved portion of input data.

mfisherlevine commented on tests/test_dark.py:
```diff
@@ -0,0 +1,170 @@
+        """
+        mean = afwMath.makeStatistics(self.exposure.getImage(), afwMath.MEAN).getValue()
+        sigma = afwMath.makeStatistics(self.exposure.getImage(), afwMath.STDEV).getValue()
+        print("5.2", mean, sigma)
```
Remove print statement.

mfisherlevine commented on tests/test_dark.py:
```diff
@@ -0,0 +1,170 @@
+                                           afwMath.STDEVCLIP, statControl).getValue()
+            # needs to be < 0.05
+            fractionalError = np.abs(sigma - amp.getReadNoise())/amp.getReadNoise())
+            self.assertLess(fractionalError, 0.71, msg=f"Test 5.3: {amp.getName()} {fractionalError}")
```
Again I'm confused by this 0.71 but that probably means I'm just not getting something and it's all OK.

mfisherlevine commented:
General comment: if (when) the sections in that technote change it's really going to suck.

mfisherlevine commented on tests/test_dark.py:
```diff
@@ -0,0 +1,170 @@
+
+            # needs to be < 0.05
+            fractionalError = np.abs(sigma - sigmaClip)/sigmaClip
+            self.assertLess(fractionalError, 3.0, msg=f"Test 5.4: {amp.getName()} {fractionalError}")
```
Again, 3?

mfisherlevine commented on tests/test_dark.py:
```diff
@@ -0,0 +1,170 @@
+        -----
+        DMTN-101 5.4:
+
+        Run a CR rejection on the result and confirm that the
+        unclipped standard deviation is consistent with the 5-sigma
+        clipped value.
```
Again, at what level?

mfisherlevine commented on tests/test_defects.py:
```diff
@@ -0,0 +1,111 @@
+        -----
+        DMTN-101 3.2
+
+        For each master frame, confirm that the median level is 1.0 to
```
What does "each" mean here? There is no frame for defects (right?!), and if this is referring to the frame-like master calibs then 1) why is it here in the defect testing, and 2) do we really expect the median of the bias frame to be 1?

mfisherlevine replied:
Hhaving said that, why do we expect _any_ of them to have a median of 1? I'd expect the darks to be scaled to an exposure _time_ of 1, but I can't see why the level would be 1.

mfisherlevine commented on tests/test_flat.py:
```diff
@@ -0,0 +1,117 @@
+        self.config.doCrosstalk = False
+        self.config.doWidenSaturationTrails = False
+        self.config.doBrighterFatter = False
+        self.config.doDefect = False
```
Why wouldn't we do defects on the flat?

mfisherlevine commented on tests/test_flat.py:
```diff
@@ -0,0 +1,117 @@
+        self.config.doUseAtmosphereTransmission = False
+
+        self.isrTask = ipIsr.IsrTask(config=self.config)
+        # This is not an independent frame.
```
As previous.

mfisherlevine commented on tests/test_flat.py:
```diff
@@ -0,0 +1,117 @@
+        """
+        mean = afwMath.makeStatistics(self.exposure.getImage(), afwMath.MEAN).getValue()
+        sigma = afwMath.makeStatistics(self.exposure.getImage(), afwMath.STDEV).getValue()
+        expectMean = 8750
+        expectSigmaMax = 3000
```
I guess these are hardcoded for the particular exposure (though sigma very seems high, no?). Will these need changing here by hand when we change data? Not arguing that that's a bad thing - I think we tend towards overengineering stuff tbh, and doing that here feels reasonable we'll change datasets quite rarely, and always by hand with human intervention, just checking that this is what's happening here.

czwa replied:
It will need changing.  I think once we've defined the tests properly, and updated DMTN-101, and find data that we can use to build calibrations that satisfy that, we'll want to move the set of test exposures and comparison values (to identify changes that may not trigger the DMTN-101 threshold to fail) into the testdata package.  Probably in the manifest.yaml file.

mfisherlevine commented on tests/test_ptc.py:
```diff
@@ -0,0 +1,58 @@
+        -----
+        DMTN-101 6.1, 6.2:
+
+        Form the function (I$_{\text{1}}$ -
+        I$_{\text{2}}$)/(I$_{\text{1}}$ + I$_{\text{2}}$) whose mean
```
I think I wrote this into the stack at some point. When this ends up being implemented give me a shout and I'll find it.

czwa replied:
I think this requires additional dev work on ip_isr and cp_pipe.  I think the "minimize discontinuities" suggests reading the PTC output gains, and applying them instead of the flat in ip_isr.  For LATISS, I think the flat scale parameters are also directly related to the correct gains as well.

mfisherlevine replied:
We should talk about this offline sometime, I fear it goes deeper than you'd naively think it should... (though I agree with everything you've said, I just know of a scary thing you might not have seen).

czwa committed:
Simplify SConstruct.

Make the DATA target a real subdirectory to build with its own
SConscript.  This additionally solves the race condition with the
tests build.

Fix the tests to use a classmethod to setup the constant ISR processed
exposure.

Fix memory error.

czwa commented:
I've updated the code here, and in `ci_cpp_gen3` to use an approach more in line with other packages.  The standard `sconsUtils` code is used in `SConstruct`, with an added target for the `DATA` directory.  The updated build structure properly orders all of the `ci_cpp` operations, as well as the tests upon completion of the `DATA` build.  I plan on adding documentation for adding new `ci_cpp` targets and squashing the unnecessary commits, but as the package is now working cleanly as intended, I'll likely merge unless there are major objections.

czwa committed:
Add documentation for new targets.
